File: JavaVersion/app/src/main/java/com/thesis/javaversion/MovieFormActivity.java
package com.thesis.javaversion;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;

import com.thesis.javaversion.database.RoomDB;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class MovieFormActivity extends AppCompatActivity {

    private TextView genreView;
    private Spinner ageRatedSpinner;
    private Button backBtn, submitBtn;
    private RoomDB database;

    private boolean[] selectedGenres;
    private ArrayList<Integer> selectedGenreList;
    private String[] genreArray;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie_form);

        init();

        submitBtn.setOnClickListener(submitMovieFormListener);

        backBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                startActivity(intent);
            }
        });

        genreView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showGenreDialog();
            }
        });
    }

    public void init() {
        Context context = getApplicationContext();
        database = RoomDB.getInstance(context);

        backBtn = findViewById(R.id.back_btn);
        submitBtn = findViewById(R.id.submit_btn);

        // Spinner age rate
        ageRatedSpinner = (Spinner) findViewById(R.id.spinner_age_rated);
        ArrayAdapter<String> ageRateAdapter = new ArrayAdapter<>(context,
                android.R.layout.simple_spinner_item,
                getResources().getStringArray(R.array.age_rate));
        ageRateAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        ageRatedSpinner.setAdapter(ageRateAdapter);

        // Checkbox Genre
        genreArray = getResources().getStringArray(R.array.genre);
        selectedGenres = new boolean[genreArray.length];
        selectedGenreList = new ArrayList<>();
        genreView = findViewById(R.id.input_genre);
    }

    private void showGenreDialog() {
        AlertDialog.Builder genreBuilder = new AlertDialog.Builder(this);
        genreBuilder.setTitle("Choose Genre(s)");

        genreBuilder.setMultiChoiceItems(genreArray, selectedGenres, new DialogInterface.OnMultiChoiceClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i, boolean checked) {
                if (checked) {
                    selectedGenreList.add(i);
                    Collections.sort(selectedGenreList);
                } else {
                    selectedGenreList.remove(i);
                }
            }
        });

        genreBuilder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                StringBuilder stringBuilder = new StringBuilder();
                for (int j = 0; j < selectedGenreList.size(); j++) {

                    stringBuilder.append(genreArray[selectedGenreList.get(j)]);

                    if (j != selectedGenreList.size() - 1) {
                        stringBuilder.append(", ");
                    }
                }
                genreView.setText(stringBuilder.toString());
            }
        });

        genreBuilder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                dialogInterface.dismiss();
            }
        });

        genreBuilder.setNeutralButton("Clear all", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                for (int j = 0; j < selectedGenreList.size(); j++) {
                    selectedGenres[j] = false;
                }
                selectedGenreList.clear();
                genreView.setText("");
            }
        });

        AlertDialog dialog = genreBuilder.create();
        dialog.setCanceledOnTouchOutside(false);
        dialog.show();
    }

    private View.OnClickListener submitMovieFormListener = new View.OnClickListener() {

        @Override
        public void onClick(View view) {
            Movie newMovie = new Movie();
            ArrayList<Boolean> isValid = new ArrayList<>();

            TextView promp = findViewById(R.id.promp_input_title);
            EditText input = findViewById(R.id.input_title);
            String text = input.getText().toString();
            isValid.add(checkInput(promp, text));
            newMovie.setTitle(text);

            promp = findViewById(R.id.promp_input_date);
            input = findViewById(R.id.input_release_date);
            text = input.getText().toString();
            isValid.add(checkInput(promp, text));
            newMovie.setReleaseDate(text);

            promp = findViewById(R.id.promp_input_age_rated);
            text = ageRatedSpinner.getSelectedItem().toString();
            isValid.add(checkAgeRateInput(promp, text));
            newMovie.setAgeRated(text);

            promp = findViewById(R.id.promp_input_genre);
            text = genreView.getText().toString();
            isValid.add(checkInput(promp, text));
            newMovie.setGenre(new ArrayList<String>(Arrays.asList(text.split(", "))));

            promp = findViewById(R.id.promp_input_score);
            input = findViewById(R.id.input_score);
            text = input.getText().toString();
            isValid.add(validScore(promp, text, newMovie));

            input = findViewById(R.id.input_img_url);
            text = input.getText().toString();
            if(Utils.IsValidUrl(text)) {
                newMovie.setImgUrl(text);
            } else {
                newMovie.setImgUrl(String.valueOf(R.drawable.no_img));
            }

            promp = findViewById(R.id.promp_input_plot);
            input = findViewById(R.id.input_plot);
            text = input.getText().toString();
            isValid.add(checkInput(promp, text));
            newMovie.setPlot(text);

            if (!isValid.contains(false)) {
              database.movieDao().insert(newMovie);
              Intent intent = new Intent(getApplicationContext(), MainActivity.class);
              startActivity(intent);
            }
        }
    };

    private boolean validScore(TextView promp, String text, Movie movie) {
        try {
            int score = Integer.parseInt(text);

            if (score > 0 && score <= 10) {
                promp.setVisibility(View.GONE);
                movie.setScore(score);
                return true;
            }

            promp.setVisibility(View.VISIBLE);
            return false;
        } catch (NumberFormatException nfe) {
            promp.setVisibility(View.VISIBLE);
            return false;
        }
    }

    private boolean checkInput(TextView promp, String text) {
        if (text.equals("")) {
            promp.setVisibility(View.VISIBLE);
            return false;
        } else {
            promp.setVisibility(View.GONE);
            return true;
        }
    }

    private boolean checkAgeRateInput(TextView promp, String text) {
        if (text.equals("Age rate *")) {
            promp.setVisibility(View.VISIBLE);
            return false;
        } else {
            promp.setVisibility(View.GONE);
            return true;
        }
    }
}

File: KotlinVersion/app/src/main/java/com/thesis/kotlinversion/MovieFormActivity.kt
package com.thesis.kotlinversion

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.thesis.kotlinversion.Utils.isValidUrl
import com.thesis.kotlinversion.database.RoomDB
import com.thesis.kotlinversion.database.RoomDB.Companion.getInstance
import java.util.*

class MovieFormActivity : AppCompatActivity() {
    private var genreView: TextView? = null
    private var ageRatedSpinner: Spinner? = null
    private var backBtn: Button? = null
    private var submitBtn: Button? = null
    private var database: RoomDB? = null
    private lateinit var selectedGenres: BooleanArray
    private var selectedGenreList: MutableList<Int>? = null
    private lateinit var genreArray: Array<String>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movie_form)
        init()
        submitBtn!!.setOnClickListener(submitMovieFormListener)
        backBtn!!.setOnClickListener { view: View? ->
            val intent = Intent(applicationContext, MainActivity::class.java)
            startActivity(intent)
        }
        genreView!!.setOnClickListener { v: View? -> showGenreDialog() }
    }

    private fun init() {
        val context = applicationContext
        database = getInstance(context)
        backBtn = findViewById(R.id.back_btn)
        submitBtn = findViewById(R.id.submit_btn)

        // Spinner age rate
        ageRatedSpinner = findViewById<View>(R.id.spinner_age_rated) as Spinner
        val ageRateAdapter = ArrayAdapter(
            context,
            android.R.layout.simple_spinner_item,
            resources.getStringArray(R.array.age_rate)
        )
        ageRateAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        ageRatedSpinner!!.adapter = ageRateAdapter

        // Checkbox Genre
        genreArray = resources.getStringArray(R.array.genre)
        selectedGenres = BooleanArray(genreArray.size)
        selectedGenreList = ArrayList()
        genreView = findViewById(R.id.input_genre)
    }

    private fun showGenreDialog() {
        val genreBuilder = AlertDialog.Builder(this)
        genreBuilder.setTitle("Choose Genre(s)")
        genreBuilder.setMultiChoiceItems(genreArray, selectedGenres)
        { dialogInterface: DialogInterface?, i: Int, checked: Boolean ->
            if (checked) {
                selectedGenreList!!.add(i)
                selectedGenreList!!.sort()
            } else selectedGenreList!!.removeAt(i)

        }
        genreBuilder.setPositiveButton("OK") { dialogInterface: DialogInterface?, i: Int ->
            val stringBuilder = StringBuilder()
            for (j in selectedGenreList!!.indices) {
                stringBuilder.append(genreArray[selectedGenreList!![j]])

                if (j != selectedGenreList!!.size - 1) stringBuilder.append(", ")
            }
            genreView!!.text = stringBuilder.toString()
        }
        genreBuilder.setNegativeButton("Cancel") { dialogInterface: DialogInterface, i: Int -> dialogInterface.dismiss() }
        genreBuilder.setNeutralButton("Clear all") { dialogInterface: DialogInterface?, i: Int ->

            for (j in selectedGenreList!!.indices) selectedGenres[j] = false

            selectedGenreList!!.clear()
            genreView!!.text = ""
        }
        val dialog = genreBuilder.create()
        dialog.setCanceledOnTouchOutside(false)
        dialog.show()
    }

    private val submitMovieFormListener = View.OnClickListener {
        val newMovie = Movie()
        val isValid = ArrayList<Boolean>()
        var promp = findViewById<TextView>(R.id.promp_input_title)
        var input = findViewById<EditText>(R.id.input_title)
        var text = input.text.toString()
        isValid.add(checkInput(promp, text, "textInput"))
        newMovie.title = text
        promp = findViewById(R.id.promp_input_date)
        input = findViewById(R.id.input_release_date)
        text = input.text.toString()
        isValid.add(checkInput(promp, text, "yearInput"))
        newMovie.releaseDate = text
        promp = findViewById(R.id.promp_input_age_rated)
        text = ageRatedSpinner!!.selectedItem.toString()
        isValid.add(checkInput(promp, text, "ageInput"))
        newMovie.ageRated = text
        promp = findViewById(R.id.promp_input_genre)
        text = genreView!!.text.toString()
        isValid.add(checkInput(promp, text, "textInput"))
        newMovie.genre = ArrayList(Arrays.asList(*text.split(", ").toTypedArray()))
        promp = findViewById(R.id.promp_input_score)
        input = findViewById(R.id.input_score)
        text = input.text.toString()
        isValid.add(checkInput(promp, text, newMovie))
        input = findViewById(R.id.input_img_url)
        text = input.text.toString()
        if (isValidUrl(text)) newMovie.imgUrl = text else newMovie.imgUrl =
            R.drawable.no_img.toString()

        promp = findViewById(R.id.promp_input_plot)
        input = findViewById(R.id.input_plot)
        text = input.text.toString()
        isValid.add(checkInput(promp, text, "textInput"))
        newMovie.plot = text
        if (!isValid.contains(false)) {
            database!!.movieDao()!!.insert(newMovie)
            val intent = Intent(applicationContext, MainActivity::class.java)
            startActivity(intent)
        }
    }

    private fun checkInput(promp: TextView, text: String, movie: Movie): Boolean {
        return try {
            val score = text.toInt()
            if (score in 1..10) {
                promp.visibility = View.GONE
                movie.score = score
                return true
            }
            promp.visibility = View.VISIBLE
            false
        } catch (nfe: NumberFormatException) {
            promp.visibility = View.VISIBLE
            false
        }
    }

    private fun checkInput(promp: TextView, text: String, type: String): Boolean {
        when (type) {
            "textInput" -> {
                return if (text == "") {
                    promp.visibility = View.VISIBLE
                    false
                } else {
                    promp.visibility = View.GONE
                    true
                }
            }
            "yearInput" -> {
                return try {
                    val movieCreatedYear = text.toInt()
                    if (movieCreatedYear in 1871..2050) {
                        promp.visibility = View.GONE
                        return true
                    }
                    promp.visibility = View.VISIBLE
                    false
                } catch (nfe: NumberFormatException) {
                    promp.visibility = View.VISIBLE
                    false
                }
            }
            "ageInput" -> {
                return if (text == "Age rate *") {
                    promp.visibility = View.VISIBLE
                    false
                } else {
                    promp.visibility = View.GONE
                    true
                }
            }
            else -> return false
        }
    }


}